Lib/tarfile.py

```diff
@@ -1432,55 +1435,66 @@ def _proc_pax(self, tarfile):
         else:
             pax_headers = tarfile.pax_headers.copy()
 
-        # Check if the pax header contains a hdrcharset field. This tells us
-        # the encoding of the path, linkpath, uname and gname fields. Normally,
-        # these fields are UTF-8 encoded but since POSIX.1-2008 tar
-        # implementations are allowed to store them as raw binary strings if
-        # the translation to UTF-8 fails.
-        match = re.search(br"\d+ hdrcharset=([^\n]+)\n", buf)
-        if match is not None:
-            pax_headers["hdrcharset"] = match.group(1).decode("utf-8")
-
-        # For the time being, we don't care about anything other than "BINARY".
-        # The only other value that is currently allowed by the standard is
-        # "ISO-IR 10646 2000 UTF-8" in other words UTF-8.
-        hdrcharset = pax_headers.get("hdrcharset")
-        if hdrcharset == "BINARY":
-            encoding = tarfile.encoding
-        else:
-            encoding = "utf-8"
-
         # Parse pax header information. A record looks like that:
         # "%d %s=%s\n" % (length, keyword, value). length is the size
         # of the complete record including the length field itself and
-        # the newline. keyword and value are both UTF-8 encoded strings.
-        regex = re.compile(br"(\d+) ([^=]+)=")
+        # the newline.
         pos = 0
-        while match := regex.match(buf, pos):
-            length, keyword = match.groups()
-            length = int(length)
-            if length == 0:
+        encoding = "utf-8"
+        raw_headers = []
+        while len(buf) > pos and buf[pos] != 0x00:
+            if not (match := _header_length_prefix_re.match(buf, pos)):
+                raise InvalidHeaderError("invalid header")
+            try:
+                length = int(match.group(1))
+            except ValueError:
+                raise InvalidHeaderError("invalid header")
+            # Headers must be at least 3 bytes, one for each of keyword, '=', and '\n'.
+            # Value is allowed to be empty.
+            if length < 3:
+                raise InvalidHeaderError("invalid header")
+            if pos + length > len(buf):
+                raise InvalidHeaderError("invalid header")
+
+            keyword_and_value = buf[match.end(1) + 1:match.start(1) + length - 1]
+            raw_keyword, equals, raw_value = keyword_and_value.partition(b"=")
+
+            # Check the framing of the header. The last character must be '\n' (0x0A)
+            if not raw_keyword or equals != b"=" or buf[match.start(1) + length - 1] != 0x0A:
                 raise InvalidHeaderError("invalid header")
-            value = buf[match.end(2) + 1:match.start(1) + length - 1]
+            raw_headers.append((length, raw_keyword, raw_value))
+
+            # Check if the pax header contains a hdrcharset field. This tells us
```

Comment:
What should we do if there are multiple hdrcharset= values in the extended pax header?  This loop combined with the one below will cause us to obey the first one that is set to BINARY, ignoring additional values for the PAX_NAME_FIELDS decoding that happens later. But the value ultimately exposed in our .pax_headers dict attribute will be that of the last one.

The https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_03 spec doesn't cover the case of multiple identically named fields being present. It makes sense to be an error to me... but matching what other tar/pax implementations do when reading a file with multiple identically named headers would be good to avoid files being interpreted differently by different tooling.

if nothing else, lets at least ensure we ignore future hdrcharset= fields within this header after we've used the value of one set to BINARY so that the value in our pax_headers dict winds up matching BINARY as that is how we behaved.

Reply:
I addressed your comments in https://github.com/python/cpython/pull/121286/commits/6ca6329d1cca1dd25cd6ab8cf2a498bbe2a9154e, I went with the simpler approach of preserving previous behavior instead of doing a survey of existing tar tools. Hope this is okay for this fix?